content/post/asdf.md

```diff
@@ -0,0 +1,240 @@
+---
+author: vadim
+date: 2018-11-10
+title: "Advanced Scientific Data Format"
+image: /post/asdf/logo.png
+description: "What is ASDF, why it is awesome, why you should and how to use it. ASDF in source{d} projects."
+categories: ["technical"]
+---
+
+There is a project we are developing at source{d} named [Modelforge](https://github.com/src-d/modelforge).
+Its goal is to abstract the serialization and the retrieval of the machine learning
+models from the users. That is, it solves the following two major problems
+every ML engineer hits:
+
+1. What to save a trained model and load it back.
+2. How to reference and distribute trained models.
+
+Let's not go offtopic with discussing (2) - there are many funny subproblems there
+which deserve a dedicated blog post about Modelforge. Instead, the focus will be on
+serialization. The Python crowd is used to solve (1) with
+[`pickle`](https://docs.python.org/3/library/pickle.html) - the built-in
+serialization module. It works nicely while these three conditions hold:
+
+1. The origin of the files is trusted and secure.
+2. The interoperability with other languages is not needed.
+3. The serialized objects are not big.
+
+As some know very well, `pickle` contains a full-featured virtual machine inside,
+which is interpreted to recreate the serialized Python objects. The advantage
+is that we can serialize literally any object, except for few rare types
+like mutexes or executable code. The implied drawbacks are substantial though:
+we should not load a pickle from unreliable source because it is vulnerable
+to remote code execution attacks; the loader must implement a virtual machine,
+which is hard with other programming languages; the memory consumption
+during serialization and deserialization can grow far beyond the original object size.
+The latter is important for machine learning and data science in particular:
+the objects can grow very big, say, bigger than a few hundred megabytes and
+pickling them is very slow and very demanding for the free operating memory.
+We witnessed 2 and even 3 times bigger memory consumption during
+`pickle`-ing which led to funny situtations when one has successfully computed
+the data but cannot dump it on disk so the whole effort becomes useless.
```

Comment:
the resulting data, but the whole effort goes to waste when there's not enough RAM for the pickling process to take place.